spring-tx/src/main/kotlin/org/springframework/transaction/reactive/TransactionalOperatorExtensions.kt

```diff
@@ -1,31 +1,56 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package org.springframework.transaction.reactive
 
-import java.util.Optional
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.reactive.asFlow
 import kotlinx.coroutines.reactive.awaitLast
 import kotlinx.coroutines.reactor.asFlux
 import kotlinx.coroutines.reactor.mono
 import org.springframework.transaction.ReactiveTransaction
+import java.util.Optional
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.coroutineContext
 
 /**
  * Coroutines variant of [TransactionalOperator.transactional] as a [Flow] extension.
  *
  * @author Sebastien Deleuze
  * @since 5.2
  */
-fun <T : Any> Flow<T>.transactional(operator: TransactionalOperator): Flow<T> =
-		operator.transactional(asFlux()).asFlow()
+fun <T : Any> Flow<T>.transactional(
+	operator: TransactionalOperator,
+	context: CoroutineContext = EmptyCoroutineContext
+): Flow<T> =
+	operator.transactional(asFlux(context)).asFlow()
 
 /**
-* Coroutines variant of [TransactionalOperator.execute] with a suspending lambda
-* parameter.
-*
-* @author Sebastien Deleuze
-* @author Mark Paluch
-* @since 5.2
-*/
-suspend fun <T : Any> TransactionalOperator.executeAndAwait(f: suspend (ReactiveTransaction) -> T?): T? =
-		execute { status -> mono(Dispatchers.Unconfined) { f(status) } }.map { value -> Optional.of(value) }
-				.defaultIfEmpty(Optional.empty()).awaitLast().orElse(null)
+ * Coroutines variant of [TransactionalOperator.execute] with a suspending lambda
+ * parameter.
+ *
+ * @author Sebastien Deleuze
+ * @author Mark Paluch
+ * @since 5.2
+ */
+suspend fun <T> TransactionalOperator.executeAndAwait(f: suspend (ReactiveTransaction) -> T): T {
```

Comment:
I am not sure about the principle of this change of null-safety, and in any case it should not be part of this PR.

@mp911de Any take on the change of null-safety requested here?

Reply:
A reactive transaction can complete without emitting a value, so we should keep `T` nullable.

Reply:
I ran into this as well. I was expecting the signature to be

 `suspend fun <T> TransactionalOperator.executeAndAwait(f: suspend (ReactiveTransaction) -> T): T {`

 i.e. for `T` to not be nullable and to not have the upper bound on `Any`.

This would make all of the following possible (and for the first scenario to not require a `!!`).

```kotlin
    val test1: Int = transactionalOperator.executeAndAwait { 1 }
    val test2: Int? = transactionalOperator.executeAndAwait { "1".toIntOrNull() }
    val test3: Int? = transactionalOperator.executeAndAwait { null }
```

Am I missing something?

Reply:
@mp911de Sorry that I didn't give a feedback for this comment for so long.
I ran into the same issue that @jamesbassett has just pointed out.

Any special reason to return `T?` instead of just `T`?
I got to always cast the returned value with `!!` even if the lambda is guaranteed to return a non-null value.

Reply:
Fixed by #29919